### ISSUE TITLE
[BUG]: crash/assertion fails when parameter is copied

### ISSUE THREAD

**Issue by HJLebbink** [modularml/mojo#1537]
### Bug description

The included minimal program crashed with the following trace. This bug is a show-stopper, and I cannot circumvent it. The crash is easy to prevent by not calling the `dummy_function`, see inline note.

```
mojo crash2.mojo
mojo: /__w/modular/modular/KGEN/lib/Interpreter/InterpreterInterface.cpp:432: auto M::InterpreterState::internalizeMemory(MutableArrayRef<mlir::Attribute>)::(anonymous class)::operator()(M::MemorySpaceAttr) const: Assertion `interned->isOwned() && "const memory cannot have pointers"' failed.
[1098048:1098048:20231221,143657.172338:ERROR file_io_posix.cc:144] open /sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq: No such file or directory (2)
[1098048:1098048:20231221,143657.172400:ERROR file_io_posix.cc:144] open /sys/devices/system/cpu/cpu0/cpufreq/scaling_max_freq: No such file or directory (2)
Please submit a bug report to https://github.com/modularml/mojo/issues and include the crash backtrace along with all the relevant source codes.
 #0 0x000055ad4cf084a7 (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x6114a7)
 #1 0x000055ad4cf0607e (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x60f07e)
 #2 0x000055ad4cf08b7f (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x611b7f)
 #3 0x00007f75dce99520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007f75dceed9fc __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x00007f75dceed9fc __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #6 0x00007f75dceed9fc pthread_kill ./nptl/pthread_kill.c:89:10
 #7 0x00007f75dce99476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x00007f75dce7f7f3 abort ./stdlib/abort.c:81:7
 #9 0x00007f75dce7f71b _nl_load_domain ./intl/loadmsgcat.c:1177:9
#10 0x00007f75dce90e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
#11 0x000055ad4eccbb16 (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x23d4b16)
#12 0x000055ad4d4b3ed0 (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0xbbced0)
#13 0x000055ad4d4b558a (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0xbbe58a)
#14 0x000055ad4ebc869c (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x22d169c)
#15 0x000055ad4d4b4fc0 (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0xbbdfc0)
#16 0x000055ad4d4b3ef7 (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0xbbcef7)
#17 0x000055ad4d4b558a (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0xbbe58a)
#18 0x000055ad4ebc869c (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x22d169c)
#19 0x000055ad4d4b4fc0 (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0xbbdfc0)
#20 0x000055ad4d4b3ef7 (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0xbbcef7)
#21 0x000055ad4ecc5daa (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x23cedaa)
#22 0x000055ad4ec10e26 (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x2319e26)
#23 0x000055ad4ec02513 (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x230b513)
#24 0x000055ad4ecc74c2 (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x23d04c2)
#25 0x000055ad4ecc6dac (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x23cfdac)
#26 0x000055ad4e4bd1fe (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x1bc61fe)
#27 0x000055ad4e4af554 (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x1bb8554)
#28 0x000055ad4e4afaef (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x1bb8aef)
#29 0x000055ad4e4a7015 (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x1bb0015)
#30 0x000055ad4e4a36c7 (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x1bac6c7)
#31 0x000055ad4ed00543 (/home/henk/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x2409543)
#32 0x00007f75dd17d253 (/lib/x86_64-linux-gnu/libstdc++.so.6+0xdc253)
#33 0x00007f75dceebac3 start_thread ./nptl/pthread_create.c:442:8
#34 0x00007f75dcf7d660 ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:83:0
Aborted
```

### Steps to reproduce

```python
fn main():
    let data = SIMD[DType.uint16, 16](0)
    alias sd1 = swap_data()
    alias sd = dummy_function(sd1) # crash here!
    #alias sd = sd1 # NOTE: no crash if you were to just use an alias

    let d2 = swap_nX[sd.get(0)](data)
    print("after:  " + str(d2))


@always_inline
fn dummy_function(sd1: SwapData) -> SwapData:
    return sd1


fn swap_data() -> SwapData:
    var result = SwapData()
    result.add(VariadicList((0,2),(1,3),(4,6),(5,7)))
    return result


@always_inline
fn swap_nX[swaps: Layer](v: SIMD[DType.uint16, 16]) -> SIMD[DType.uint16, 16]:
    fn gen_perm[swaps: Layer]() -> StaticIntTuple[16]:
        var result = StaticIntTuple[16]()
        for i in range(16):
            result[i] = i

        for i in range(len(swaps)):
            let from_ = swaps.get_min(i)
            let to_ = swaps.get_max(i)

            let tmp = result[to_]
            result[to_] = result[from_]
            result[from_] = tmp

        return result

    alias permutations = gen_perm[swaps]()
    return v # just do nothing, but removing gen_perm optimized the  


struct SwapData:
    var data: DynamicVector[Layer]

    fn __init__(inout self):
        self.data = DynamicVector[Layer]()

    fn __copyinit__(inout self: Self, existing: Self):
        self.data = existing.data

    # get the i-th layer
    fn get(self: Self, idx: Int) -> Layer:
        if (idx >= 0) & (idx < len(self.data)):
            return self.data[idx]
        else:
            print("ERROR: SwapData.get")
            return Layer()


    # add a layer of swaps
    fn add(inout self, layer: VariadicList[Tuple[Int, Int]]):
        self.data.push_back(Layer(layer))


struct Layer(CollectionElement, Sized):
    var data: DynamicVector[Int]

    @staticmethod
    fn pack(t: Tuple[Int, Int]) -> Int:
        let v1 = t.get[0, Int]()
        let v2 = t.get[1, Int]()
        if v1 < v2:
            return (v1 << 16) | v2
        else:
            return (v2 << 16) | v1

    # unpack Tuple[Minimum, Maximum]
    @staticmethod
    fn unpack(v: Int) -> Tuple[Int, Int]:
        return (v >> 16, v & 0xFFFF)

    fn __init__(inout self):
        self.data = DynamicVector[Int]()

    fn __init__(inout self, v: VariadicList[Tuple[Int, Int]]):
        self.data = DynamicVector[Int](v.__len__())
        for i in range(v.__len__()):
            self.data.push_back(Layer.pack(v[i]))

    # trait CollectionElement
    fn __copyinit__(inout self, existing: Self):
        self.data = existing.data

    # trait CollectionElement
    fn __moveinit__(inout self, owned existing: Self):
        self.data = existing.data ^

    # trait CollectionElement
    fn __del__(owned self: Self):
        pass

    # trait Sized
    fn __len__(self) -> Int:
        return len(self.data)

    fn get_min(self, idx: Int) -> Int:
        return Layer.unpack(self.data[idx]).get[0, Int]()

    fn get_max(self, idx: Int) -> Int:
        return Layer.unpack(self.data[idx]).get[1, Int]()
```

### System information

```shell
$ wsl --version
WSL version: 2.0.9.0
Kernel version: 5.15.133.1-1
WSLg version: 1.0.59
MSRDC version: 1.2.4677
Direct3D version: 1.611.1-81528511
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.22631.2715

$ mojo --version
mojo 0.6.1 (876ded2e)

$ modular --version
modular 0.2.2 (95d42445)
```

**Comment by HJLebbink** [comment]
A way to circumvent the issue is by rewriting `dummy_function` to:
```python
@always_inline
fn dummy_function[sd1: SwapData]() -> SwapData:
    return sd1
```

**Comment by Mogball** [comment]
@HJLebbink could you possibly provide a smaller repro! Thank you so much for filing

**Comment by HJLebbink** [comment]
Here is a smaller repro: maybe related: maybe related: https://github.com/modularml/mojo/issues/1606

```python
fn main():
    let data = SIMD[DType.uint16, 16](0)
    alias SD1 = swap_data()
    let d2 = xyzzy[SD1[0]](data)
    print(d2)


fn swap_data() -> DynamicVector[Layer]:
    var result = DynamicVector[Layer]()
    result.push_back(VariadicList((0,2),(1,3),(4,6),(5,7)))
    return result


fn xyzzy[swaps: Layer](v: SIMD[DType.uint16, 16]) -> SIMD[DType.uint16, 16]:
    fn gen_perm[swaps: Layer]() -> StaticIntTuple[16]:
        let result = StaticIntTuple[16]()
        for i in range(len(swaps.data)):
            let from_ = swaps.data[i] # removing this line removes the crash
        return result

    alias permutations = gen_perm[swaps]() # changing alias to let removes the crash
    return v # just do nothing, removing gen_perm removes the crash


struct Layer(CollectionElement):
    var data: DynamicVector[Int]

    fn __init__(inout self, v: VariadicList[Tuple[Int, Int]]):
        self.data = DynamicVector[Int]()
        for i in range(v.__len__()):
            self.data.push_back(v[i].get[0, Int]())

    fn __copyinit__(inout self, existing: Self):
        self.data = existing.data

    fn __moveinit__(inout self, owned existing: Self):
        self.data = existing.data ^

    fn __del__(owned self: Self):
        pass
```

**Comment by Mogball** [comment]
Thanks! We'll take a look

**Comment by HJLebbink** [comment]
Update: still an issue in mojo 0.6.3 (f58f8b94)

**Comment by HJLebbink** [comment]
Update: still an issue in mojo 0.7.0 (af002202)

**Comment by Mogball** [comment]
Seems to be fixed at top-of-tree. The program now outputs

```
mojo /tmp/test.mojo
after:  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```

Not sure if this is correct, however.

**Comment by HJLebbink** [comment]
@Mogball any estimate in which version this will be fixed: mojo 0.7.1 (af002202) still has it.

**Comment by Mogball** [comment]
In the next release
